app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java

```diff
@@ -86,21 +104,67 @@ boolean checkExists(final Contact contact) throws IOException {
     }
 
     /**
-     * Deserializes a contact through a byte array and sends it to the database.
+     * The main method that deals with the result of the scocket stream
+     * and dispatches the data approprietly in order to parse it.
      *
-     * @param bytes byte array representing a contact
-     * @return Whether or not the process finished successfully
+     * @param bytes Data represented as bytes (usually received from the socket)
      */
-    Contact saveContact(final byte[] bytes) {
-        LOGGER.info("Saving these bytes: %s", bytes);
-        final Contact contact;
+    @SuppressWarnings({"checkstyle:cyclomaticcomplexity", "PMD.StdCyclomaticComplexity", "PMD.CyclomaticComplexity"})
+    void parseInput(final byte[] bytes) {
+        LOGGER.info("Reading these bytes: %s", bytes);
+        final DataWrapper dataWrapper;
         try (ByteArrayInputStream bis = new ByteArrayInputStream(bytes);
              ObjectInputStream ois = new ObjectInputStream(bis)) {
-            contact = (Contact) ois.readObject();
+            dataWrapper = (DataWrapper) ois.readObject();
         } catch (final ClassNotFoundException | IOException e) {
             LOGGER.error(" Couldn't start deserialization!", e);
             throw (DeserializationException) new DeserializationException("The contact could not be deserialized").initCause(e);
         }
+        switch (dataWrapper.getClazz().getSimpleName()) {
+            case CONTACT_CLASS:
+                LOGGER.info("Content was a contact");
+                saveContact((Contact) dataWrapper.getData());
+                break;
+            case "ArrayList":
+                LOGGER.info("Content was a List");
+                final ArrayList<Object> list = (ArrayList<Object>) dataWrapper.getData();
+                if (!list.isEmpty()) {
+                    for (final Object o : list) {
+                        if (o.getClass().getSimpleName().equals(CONTACT_CLASS)) {
```

Comment:
can't you also just say instanceof Contact? Don't know if this is better though.